doc: Add Google Programmable Search Engine

kartben opened the pull request:
Following discussions at ZDS, where the initial feedback was pretty positive, this PR adds Google Programmable Search engine to the documentation (while leaving it possible to easily revert to the built-in Sphinx search engine). 
As the styling of the search results is apparently not really thought to be easy to tweak, the gcs.css stylesheet will more than likely require further work.

Some noteworthy items:

* [Refinements](https://support.google.com/programmable-search/answer/12425418) for "boards", "DTS bindings", or "API" make it really easy to only search in a particular subset of the docs.
* As we are a nonprofit, we would be able to get rid of advertisments in the results page.
* As I wanted the initial changes to be minimal, I re-used our existing searchbox, but we could look at using the one provided by Google, to get autocompletion (see https://cse.google.com/cse?cx=746031aa0d56d4912#gsc.tab=0)

Fixes #55173.

Direct link to CI-built doc: https://builds.zephyrproject.io/zephyr/pr/60018/docs/

kartben committed:
doc: Add Google Programmable Search Engine

Add Google Programmable Search engine to the documentation, while
leaving it possible to easily revert to the built-in Sphinx search
engine. As the styling of the search results is apparently not thought
to be easy to tweak, the gcs.css stylesheet might require further
improvements.
Fixes #55173.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

henrikbrixandersen commented on doc/_templates/search.html:
```diff
@@ -0,0 +1,28 @@
+{% block body %}
+  {% if google_searchengine_id %}
+    <h2>{{ _('Search Results') }}</h2>
+    <script async src="https://cse.google.com/cse.js?cx={{ google_searchengine_id }}">
```
Depending on an external partner (one that may indeed be unavailable, have limited access, be biased, not allowed by some companies' infrastructure) is a no-go for our project documentation IMO. Same goes for locally built documentation.

kartben replied:
Point taken, and also one of the reasons why I'd like to keep it super straightforward to switch back to "self-contained" search. 

Regarding relying on 3rd party services, our code is hosted on GitHub, our CI runs on GitHub and AWS, our docs are hosted on AWS, and a lot of the community interactions happen on Discord, all of which can be unavailable at times, or not be accessible / blocked for some folks :/

henrikbrixandersen reviewed (dismissed):
3rd party dependencies in our documentation - especially controversial ones - are a no-go for me.

thedjnK reviewed (dismissed):
Search no longer works for anyone blocking google or google services, no search is far worse than a half baked search solution

fabiobaltieri commented:
> Search no longer works for anyone blocking google or google services, no search is far worse than a half baked search solution

Can we have automatic fallback for that case? I imagine it'd be possible to detect if the script is blocked. I find the current search to be roughly equivalent to no search at all, this would make it it way more useful for anyone who has access to it.

kartben commented:
> > Search no longer works for anyone blocking google or google services, no search is far worse than a half baked search solution
> 
> Can we have automatic fallback for that case? I imagine it'd be possible to detect if the script is blocked. I find the current search to be roughly equivalent to no search at all, this would make it it way more useful for anyone who has access to it.
> 
> I don't see why we can't have both.

Yes pretty sure implementing auto-fallback (and generally speaking, allowing customization based on the user's preference, similar to how one might decide they prefer the dark theme) should be reasonably easy.

fabiobaltieri commented:
> Yes pretty sure implementing auto-fallback (and generally speaking, allowing customization based on the user's preference, similar to how one might decide they prefer the dark theme) should be reasonably easy.

Good stuff! Talking about dark theme, I think it still needs a few minor tweaks:

![x](https://github.com/zephyrproject-rtos/zephyr/assets/891546/7c42a8b7-8ddc-4fdd-92ea-f2cfa83ab034)

kartben commented:
> Good stuff! Talking about dark theme, I think it still needs a few minor tweaks:

Yup I caught that one too, thanks! Styling of the results is incredibly painful, fwiw, and I wish there was a way to start with no CSS at all rather than the inline styles that get automatically/forcefully applied.

keith-zephyr commented:
There are still some dead links from Google search that generate bad results.  Searching on "USB-C" yields a couple links with text "If you are looking for the documentation form a previous release..."

![image](https://github.com/zephyrproject-rtos/zephyr/assets/75284258/8aabc3ba-e9d4-4f30-9331-3cb2f812d23c)

gmarull reviewed (commented):
Any improvement to doc search is a good idea, what we have now is unusable. Added one suggestion.

fabiobaltieri commented:
Convenient link if anyone else wants to try it: https://builds.zephyrproject.io/zephyr/pr/60018/docs/. I have it bookmarked and have been test driving it for the past week.

nordicjm reviewed (changes requested):
Leaving as -1 for using a third party search engine that cares not for privacy

nordicjm commented:
Has anyone reached out to sphinx about the zephyr documentation search specifically and asked if there are ways to improve it, or things sphinx they could change/improve to make it work better?

gmarull commented:
> Has anyone reached out to sphinx about the zephyr documentation search specifically and asked if there are ways to improve it, or things sphinx they could change/improve to make it work better?

It's not really a Zephyr problem, even though our documentation structure doesn't help (i.e. tons of pages about dt bindings/boards/samples). IMHO, search will only work if we switch to some server-side search engine, whether it's private (e.g. Google, Algolia like [Godot](https://docs.godotengine.org/en/latest/index.html)) or open-source.

nashif commented:
I just got to play with this and I think it is great. We should get this in ASAP. A few things we need to resolve before we go with this:
- Figure out how to remove Ads
- Allow building the docs with default search, for those hosting the docs somewhere else. Maybe enable this only in the release/publication build of the docs and not while building as a developer, ie. the default.

nashif reviewed (dismissed):
This is very nice. We need to solve this issue with documentation search and I have not seen better options. This is adding search to a project hosted documentation website, it is not changing what Zephyr is and anyone re-using the documentation can enable some other tool for search. 
If we do not go with this '3rd party' solution, we will have to do this ourselves, but for this to happen, we need someone to step up.

kartben commented:
> - Figure out how to remove Ads

I just learnt that LF is not eligible to "Google for Nonprofits" :( @keith-zephyr @fabiobaltieri to maybe help? Otherwise cost is $5 per 1,000 queries.

> - Allow building the docs with default search, for those hosting the docs somewhere else. Maybe enable this only in the release/publication build of the docs and not while building as a developer, ie. the default.

Yep, +1

nashif commented:
> I just learnt that LF is not eligible to "Google for Nonprofits"

thats bad, why do need to apply as LF? Can't we apply as the Zephyr project?

thedjnK commented:
I don't know if the documentation build has not been updated or not since it's never seemed to update after the first generation in PRs I've seen, but if this does not use the sphinx search when people do not have or want to use an incredibly dodgy third party data harvesting company then this is a non starter. As of right now, in the linked documentation in the PR, search does not work, it is completely useless for me

keith-zephyr commented:
> > * Figure out how to remove Ads
> 
> I just learnt that LF is not eligible to "Google for Nonprofits" :( @keith-zephyr @fabiobaltieri to maybe help? Otherwise cost is $5 per 1,000 queries.
> 

I've engaged with one of our partner engagement managers and we're trying to get our Google for non profits team involved.

kartben commented:
> > > * Figure out how to remove Ads
> > 
> > 
> > I just learnt that LF is not eligible to "Google for Nonprofits" :( @keith-zephyr @fabiobaltieri to maybe help? Otherwise cost is $5 per 1,000 queries.
> 
> I've engaged with one of our partner engagement managers and we're trying to get our Google for non profits team involved.

That is awesome news, thank you for sharing the update, Keith.

kartben commented:
Thank you for the update, @keith-zephyr!

kartben committed:
doc: exclude some HTML DIVs from search engine "snippets"

Some portions of the documentation do not make sense when shown as
"snippets" in search engines' results. Use "data-nosnippet" attribute to
instruct search engines (Google for sure, hopefully others in the
future) to exclude them.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: Make search engine configurable

Add UI to select preferred search engine when Google Programmable Search is
enabled. The user's preference is saved using local storage.

This also makes the search input field of type "search" for better UX (in
particular on mobile).

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: remove readthedocs-sphinx-search  CSS rules

The "Sphinx search extension" CSS rules are not needed as we are not
using this extension since we are not hosting on readthedocs.io.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben commented:
Latest update to this PR introduces a way to select the search engine, which is persisted in the local storage of the user's browser, and defaults to Sphinx built-in search functionality.

The styling of the Google results has also been improved (fixed dark theme, rendered the "refinement labels" with a more distinguishable background and rounded border).

https://builds.zephyrproject.io/zephyr/pr/60018/docs/

![image](https://github.com/zephyrproject-rtos/zephyr/assets/128251/75db047c-57a0-4fda-b4e7-654c36792c82)

![image](https://github.com/zephyrproject-rtos/zephyr/assets/128251/4e0685a9-4e17-4186-8664-1412cac656fa)

kartben commented on doc/conf.py:
```diff
@@ -169,7 +169,11 @@
+    # Set google_searchengine_id to your Search Engine ID to replace built-in search
+    # engine with Google's Programmable Search Engine.
+    # See https://programmablesearchengine.google.com/ for details.
+    "google_searchengine_id": "746031aa0d56d4912"
```
Please check out latest update to the PR -- folks can now decide in the UI which engine they want to use. I guess we can still provide some level of build-time configuration though so that search engine selection widget is available only in case folks have enabled the use of a 3rd party engine when generating the docs.

nordicjm reviewed (commented):
Seems to be a much better approach! Some queries, if I search for something with google selected it doesn't work (this is expected) what is unexpected though is the search term vanishes from the search box on the left, so if I then switch back to the working search method, I have to retype the query.

Another question is can this be completely disabled for e.g. local builds and any out of tree (i.e. downstream) users of zephyr?

kartben commented:
> There are still some dead links from Google search that generate bad results. Searching on "USB-C" yields a couple links with text "If you are looking for the documentation form a previous release..."

Rather than dead links, I think it's the way "snippets" look that can be improved, as they tend to include text that might be irrelevant, like "... if you are looking for the documentation of ..." in your example.
This commit should help improve the situation https://github.com/zephyrproject-rtos/zephyr/pull/60018/commits/772d78e8e3b1101e3e20123afb212f58f2a3d718

gmarull commented:
> Latest update to this PR introduces a way to select the search engine, which is persisted in the local storage of the user's browser, and defaults to Sphinx built-in search functionality.
> 
> The styling of the Google results has also been improved (fixed dark theme, rendered the "refinement labels" with a more distinguishable background and rounded border).
> 
> https://builds.zephyrproject.io/zephyr/pr/60018/docs/
> 
> ![image](https://user-images.githubusercontent.com/128251/281055018-75db047c-57a0-4fda-b4e7-654c36792c82.png)
> 
> ![image](https://user-images.githubusercontent.com/128251/281055077-4e0685a9-4e17-4186-8664-1412cac656fa.png)

TBH, cluttering the UI with such dropdown seems too much for me, I'll likely need to add a script to hide this on my end. Why not hide it somehow? Maybe a tiny dropdown button in the same search box?

kartben commented:
> [...] Maybe a tiny dropdown button in the same search box?

Ya I was looking into having the dropdown within the search box but eventually gave up as I wasn't really happy about how it looked and how unintuitive this was. I will probably give it another try though.

nashif commented:
Does not make sense to have the built-in be the default, I would not even go far to offer this option at this level. If we are going to be paying for google then this needs to be the default and usable and ready to go and transparent to the user. This could be a build time option to enable one or the other.

thedjnK commented:
> Does not make sense to have the built-in be the default, I would not even go far to offer this option at this level. If we are going to be paying for google then this needs to be the default and usable and ready to go and transparent to the user. This could be a build time option to enable one or the other.

Disagree, the built in search just works as it is, the google one is quite rightly blocked by some people (myself included). A search engine that half works is better than complete inability to search

kartben commented:
I think this should be good to go now. Changes since latest update from Nov. 7:

- I've made the UI to select which engine to use less intrusive. I would like to stick to a select box as this provides better usability.
- Default search engine for "latest" documentation is Google, while the Sphinx built-in search is used for old/archived versions of the docs (but people can still force Google if they feel like it, and this search engine preference will be saved independently of the one for "latest").
- Added "data-nosnippet" hint in a few places to instruct search engines to exclude irrelevant text from the snippets showing up in the query results. (#77546a59b9ca8e8fb5453d5b7941a6e920ed1413)
- When `google_searchengine_id` option is not set in conf.py, the experience is exactly as before, i.e. no UI to select a custom search engine, and only built-in search is available/used.
- CSS styling should be complete now, both for dark and light theme.

nordicjm commented on doc/_templates/searchbox.html:
```diff
@@ -0,0 +1,65 @@
+  (ex. built-in search, Google Custom Search, ...)
+#}
+
+{%- if ('singlehtml' not in builder) %}
```
Not sure what `singlehtml` refers to but will the alternative search be disabled in PR builds and builds people do on their desk? The search is only going to show information from the online source which probably won't match up with these builds (not sure if it would even give local links or if the links would all be to the live documentation site)

kartben replied:
PR builds default to "sphinx" since they don't have a URL that starts with /latest, and same for local build on your desk. You can still switch to google (and have it saved as a preference in the browser's local storage) but to your point this will likely not be what one would want to do.

BTW `singlehtml` is for the Sphinx build mode where everything ends up in one single HTML page (something I know you're very found of ;-)), and this guard is basically the same the Read The Docs theme has in place as search is not available (needed) in single page mode.

gmarull commented:
I think this too ugly:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/1cce306a-a260-44dc-a248-e8584158aad1)

Sorry, can't approve. I'd suggest exploring an in-field drop-down. As usual in Zephyr, I think we're overcomplicating things, leading to a bad UX.

kartben commented:
> Sorry, can't approve. I'd suggest exploring an in-field drop-down. As usual in Zephyr, I think we're overcomplicating things, leading to a bad UX.

@gmarull Please have another look. Ugliness should be gone know, and I took some extra care to ensure the new UI is still accessible (which is why I went with standard html elements in the first place)

gmarull reviewed (approved):
thanks for improving the UI, much better now :-)

kartben commented:
> thanks for improving the UI, much better now :-)

Thank you! I will probably make another accessibility pass in a second stage as it is still not 100% there, but this is a topic worth making a pass across the entire documentation anyway.

nordicjm reviewed (commented):
How will this work with versioned documentation, so let's assume this is merged now and we have released zephyr 3.6 and 3.7, if someone goes to the 3.6 documentation and searches, will it correctly search the 3.6 documentation and return links to 3.6 only?

kartben commented:
> How will this work with versioned documentation, so let's assume this is merged now and we have released zephyr 3.6 and 3.7, if someone goes to the 3.6 documentation and searches, will it correctly search the 3.6 documentation and return links to 3.6 only?

Great question.
When browsing "non-latest" documentation (ex. future `https://docs.zephyrproject.org/3.6.0/...` URLs), one's browser will default to built-in Sphinx search and the historical search will work as it always as.
It will still be possible for the user to force the use of Google (and the user preference is persisted in the browser local storage) if they wish, but in this case they will get "fresh" results from the Google index, since we're intentionally aiming for old (aka non-latest / non-LTS) releases to not appear in public search engines.

kartben commented:
Thanks to everyone who reviewed/approved :) I will be putting DNM to make sure this doesn't inadvertently get merged before we have a Google billing account in place and Ad-less mode properly configured

kartben commented:
This is ready to go as ad-less mode has now been enabled.

kartben commented:
@henrikbrixandersen please have another look
